processor/src/main/java/org/mapstruct/ap/internal/model/assignment/Assignment.java

```diff
@@ -42,7 +42,20 @@
         /** assignment is first mapped (builtin/custom), then the result is type converted */
         MAPPED_TYPE_CONVERTED,
         /** assignment is first type converted, and then mapped (builtin/custom) */
-        TYPE_CONVERTED_MAPPED
+        TYPE_CONVERTED_MAPPED;
+
+        public boolean isDirect() {
+            return this == DIRECT;
+        }
+
+        public boolean isTypeRelated() {
```

Comment:
What do this booleans signify? `isTypeRelated` and `isMapRelated` both concern mapping and conversion. Its just the order in which they are called that matters. So if you map a method first, you allow null (no npe check) because the implementation of the called mapping method should handle null (it could also have a meaning).

The two introduced methods obfuscate the original intention of these enum constant. Especially if you consider that `isMapRelated` could also be related to map's (they require special treatments as well in some places).